File: package/luci-app-glinet-privacy/luasrc/model/cbi/glinet_privacy/killswitch.lua
--[[
Kill switch — UCI: privacy
]]

local sys = require "luci.sys"

m = Map("privacy", translate("Privacy kill switch"),
	translate("When enabled, the watchdog drops forwarded traffic from LAN to the physical WAN if WireGuard or Tor is unhealthy. Configure interfaces to match <code>ip link</code> and your WWAN device."))

s = m:section(NamedSection, "main", "privacy", translate("Settings"))
s.addremove = false

s:option(Flag, "enabled", translate("Enable watchdog"),
	translate("If disabled, killswitch rules are flushed.")).rmempty = false

s:option(Value, "wg_if", translate("WireGuard interface"), translate("e.g. wg0")).rmempty = true
s:option(Flag, "require_wg", translate("Require WG interface UP")).rmempty = false
s:option(Flag, "require_tor", translate("Require Tor running")).rmempty = false
s:option(Value, "lan_dev", translate("LAN bridge / device"), translate("e.g. br-lan; leave empty to use network.lan.device")).rmempty = true
s:option(Value, "wan_dev", translate("Physical WAN device"), translate("e.g. wwan0 or eth0; required for correct FORWARD match")).rmempty = true

function m.on_commit(self)
	sys.call("/usr/bin/privacy-killswitch-watchdog.sh >/dev/null 2>&1")
end

return m

File: package/luci-app-glinet-privacy/htdocs/luci-static/resources/view/glinet_privacy/killswitch.js
'use strict';
'require view';
'require dom';
'require fs';
'require uci';
'require form';
'require ui';

function glpJson(sub) {
	return fs.exec('/usr/libexec/glinet-privacy/luci-json.sh', [sub]).then(function (r) {
		if (r.code !== 0)
			throw new Error(r.stderr || 'luci-json failed');
		return JSON.parse(r.stdout || '{}');
	});
}

return view.extend({
	load: function () {
		return Promise.all([uci.load('privacy'), glpJson('net_probe')]);
	},

	render: function (data) {
		var net = data[1];
		var m = new form.Map('privacy', _('Kill switch'));
		var s = m.section(form.NamedSection, 'main', _('Watchdog settings'));
		s.option(form.Flag, 'enabled', _('Kill switch watchdog enabled'));
		s.option(form.Value, 'wg_if', _('WireGuard interface name'));
		s.option(form.Flag, 'require_wg', _('Require WireGuard'));
		s.option(form.Flag, 'require_tor', _('Require Tor'));
		s.option(form.Value, 'lan_dev', _('LAN device (optional)'));
		s.option(form.Value, 'wan_dev', _('WAN device (optional)'));
		var o = s.option(form.ListValue, 'vendor_gl_vpn_killswitch', _('GL.iNet VPN kill switch sync'));
		o.value('leave', _('Leave (do not change)'));
		o.value('on', _('On'));
		o.value('off', _('Off'));

		return m.render().then(function (node) {
			var hint = '';
			if (net && net.watchdog_lan)
				hint = _('Watchdog LAN') + ': ' + net.watchdog_lan + ' · ' + _('WAN') + ': ' + (net.watchdog_wan || '—');
			var btn = E('div', { class: 'cbi-section' }, [
				E('button', {
					class: 'btn cbi-button',
					click: ui.createHandlerFn(this, function () {
						return Promise.all([
							fs.exec('/usr/libexec/glinet-privacy/apply-vendor-vpn-killswitch.sh'),
							fs.exec('/usr/bin/privacy-killswitch-watchdog.sh'),
							fs.exec('/etc/init.d/firewall', ['reload'])
						]).then(function () {
							ui.addNotification(null, E('p', {}, [_('Vendor VPN + watchdog + firewall reload done.')]));
						});
					})
				}, [_('Apply vendor VPN + watchdog')])
			]);
			return E('div', {}, [E('h2', {}, [_('Kill switch')]), E('p', {}, [hint]), node, btn]);
		});
	}
});
